Sources/Amplitude/AMPDeviceInfo.m

```diff
@@ -101,27 +101,46 @@ - (NSString *)model {
 }
 
 - (NSString *)carrier {
+    // unable to fetch carrier information
     if (!_carrier) {
-        Class CTTelephonyNetworkInfo = NSClassFromString(@"CTTelephonyNetworkInfo");
-        SEL subscriberCellularProvider = NSSelectorFromString(@"subscriberCellularProvider");
-        SEL carrierName = NSSelectorFromString(@"carrierName");
-        if (CTTelephonyNetworkInfo && subscriberCellularProvider && carrierName) {
-            networkInfo = [[NSClassFromString(@"CTTelephonyNetworkInfo") alloc] init];
-            id carrier = nil;
-            id (*imp1)(id, SEL) = (id (*)(id, SEL))[networkInfo methodForSelector:subscriberCellularProvider];
-            if (imp1) {
-                carrier = imp1(networkInfo, subscriberCellularProvider);
-            }
-            NSString *(*imp2)(id, SEL) = (NSString *(*)(id, SEL))[carrier methodForSelector:carrierName];
-            if (imp2) {
-                _carrier = imp2(carrier, carrierName);
+        _carrier = @"Unknown";
+    }
+
+    Class CTTelephonyNetworkInfo = NSClassFromString(@"CTTelephonyNetworkInfo");
+    if (!CTTelephonyNetworkInfo) {
+        return _carrier;
+    }
+
+    networkInfo = [[CTTelephonyNetworkInfo alloc] init];
+    id carrier = nil;
+    SEL carrierName = NSSelectorFromString(@"carrierName");
+
+    // subscriberCellularProvider is deprecated after iOS 12.0
+    if (@available(iOS 12, *)) {
+        SEL serviceSubscriberCellularProviders = NSSelectorFromString(@"serviceSubscriberCellularProviders");
+        id carrierMap = nil;
+        id (*imp1)(id, SEL) = (id (*)(id, SEL))[networkInfo methodForSelector:serviceSubscriberCellularProviders];
+        if (imp1) {
+            carrierMap = imp1(networkInfo, serviceSubscriberCellularProviders);
+        }
+        SEL mobileNetworkCode = NSSelectorFromString(@"mobileNetworkCode");
+        for (NSString* key in carrierMap){
+            NSString *(*getMobileNetworkCode)(id, SEL) = (NSString *(*)(id, SEL))[carrierMap[key] methodForSelector:mobileNetworkCode];
+            if (getMobileNetworkCode && getMobileNetworkCode(carrierMap[key], mobileNetworkCode) != nil){
+                carrier = carrierMap[key];
```

Comment:
Once you get the value, should we just exit the loop?

Reply:
Thanks for reviewing! I think both are good IMO, either exit the loop early so we choose the first one, or here we choose the last one. Updated to exit early to same one potential execution as you suggested.